File: services/knowledge-service/app/routers/public/ontology.py
"""KG graph-ontology public router (epic 2026-06-20, lane LC).

Fills the L1-preregistered stub with the tiered graph-schema surface:
list/read, resolved-schema read, adopt (copy-down, M1 adopt-gate), sync
(diff/apply), and per-tier child CRUD (additive + deprecate-only, M3).

Contract: contracts/api/knowledge-service/ontology.yaml.

Tenancy (CLAUDE.md › User Boundaries):
  * **System tier** is read-only over this API. Reads serve everyone; writes
    (PATCH/DELETE/CRUD) on a system row → 403; system *create* → 501 placeholder.
  * **User tier** writes are owner-scoped (`scope_id == caller`).
  * **Project tier** writes are Manage-gated via `require_project_grant(MANAGE)`
    (resolve-to-owner): the grant gate authorizes the caller and the write lands
    as the project owner. Adopt/sync-apply/CRUD all go through this gate.
  * The mutations repo additionally refuses any write on a system row, so a
    mis-routed write fails closed at the data layer too.

Spec: docs/specs/2026-06-20-knowledge-graph-customizable-ontology.md §8 M1/M3,
§10-A3/A4.
"""

from __future__ import annotations

import logging
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Path, Query, Response, status
from pydantic import BaseModel, Field

from app.auth.grant_deps import GrantLevel, require_project_grant
from app.clients.glossary_ontology_client import (
    GlossaryOntologyClient,
    HttpGlossaryOntologyClient,
)
from app.config import settings
from app.db.ontology_models import GraphSchema
from app.db.pool import get_knowledge_pool
from app.db.repositories.graph_schemas import GraphSchemasRepo
from app.db.repositories.ontology_mutations import (
    ChildNotFoundError,
    DuplicateChildError,
    NeedsGlossaryError,
    OntologyMutationsRepo,
    SchemaNotWritableError,
    SyncConflictError,
)
from app.db.repositories.projects import ProjectsRepo
from app.deps import get_grant_client, get_projects_repo
from app.middleware.jwt_auth import get_current_user
from app.ontology.glossary_gate import (
    adopt_with_autocreate_glossary,
    resolve_adopt_glossary_codes,
)

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/v1/kg",
    tags=["kg-ontology"],
    dependencies=[Depends(get_current_user)],
)

_CODE_MAX = 120
_NAME_MAX = 200


# ── DI ──────────────────────────────────────────────────────────────────────
def get_graph_schemas_repo() -> GraphSchemasRepo:
    return GraphSchemasRepo(get_knowledge_pool())


def get_ontology_mutations_repo() -> OntologyMutationsRepo:
    return OntologyMutationsRepo(get_knowledge_pool())


_glossary_ontology_client: GlossaryOntologyClient | None = None


def get_glossary_ontology_client() -> GlossaryOntologyClient:
    """Singleton KG→glossary ontology-read client (long-lived httpx). Tests
    override this dep with `FakeGlossaryOntologyClient` so the adopt-gate runs
    without a live glossary."""
    global _glossary_ontology_client
    if _glossary_ontology_client is None:
        _glossary_ontology_client = HttpGlossaryOntologyClient(
            base_url=settings.glossary_service_url,
            internal_token=settings.internal_service_token,
            timeout_s=getattr(settings, "glossary_client_timeout_s", 10.0),
        )
    return _glossary_ontology_client


# ── request / response models ────────────────────────────────────────────────
class GraphSchemaListResponse(BaseModel):
    items: list[GraphSchema]


class GraphSchemaTree(GraphSchema):
    """Full schema tree: the summary fields (inherited) + children arrays.

    Mirrors the contract `allOf [GraphSchemaSummary, {children}]` — children are
    inlined at the top level, NOT nested under a `schema` key."""

    edge_types: list[dict] = Field(default_factory=list)
    fact_types: list[dict] = Field(default_factory=list)
    vocab_sets: list[dict] = Field(default_factory=list)
    node_kinds: list[dict] = Field(default_factory=list)


class GraphSchemaPatch(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=_NAME_MAX)
    description: str | None = None
    allow_free_edges: bool | None = None


class AdoptRequest(BaseModel):
    source_schema_id: UUID
    acknowledge_optional_gaps: bool = False


class AdoptPreviewRequest(BaseModel):
    source_schema_id: UUID


class EdgeTypeCreate(BaseModel):
    code: str = Field(min_length=1, max_length=_CODE_MAX)
    label: str = Field(min_length=1)
    directed: bool = True
    source_node_kinds: list[str] = Field(default_factory=list)
    target_node_kinds: list[str] = Field(default_factory=list)
    temporal: bool = False
    provenance_required: bool = False
    cardinality: str = "multi_active"
    description: str = ""


class FactTypeCreate(BaseModel):
    code: str = Field(min_length=1, max_length=_CODE_MAX)
    label: str = Field(min_length=1)
    description: str = ""


class VocabValueCreate(BaseModel):
    code: str = Field(min_length=1, max_length=_CODE_MAX)
    label: str = Field(min_length=1)
    metadata: dict[str, Any] = Field(default_factory=dict)


class SchemaNodeKindCreate(BaseModel):
    kind_code: str = Field(min_length=1, max_length=_CODE_MAX)
    strength: str = Field(pattern="^(required|optional)$")


class SyncDecision(BaseModel):
    node_type: str
    parent_code: str | None = None
    code: str
    choice: str = Field(pattern="^(keep_mine|take_theirs)$")


class SyncApplyRequest(BaseModel):
    base_source_hash: str
    decisions: list[SyncDecision] = Field(default_factory=list)


# ── helpers ───────────────────────────────────────────────────────────────────
def _not_found() -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="not found")


def _forbidden(msg: str = "insufficient access") -> HTTPException:
    return HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=msg)


async def _writable_schema_for_caller(
    schema_id: UUID, caller: UUID, mutations: OntologyMutationsRepo, gc, projects: ProjectsRepo
) -> GraphSchema:
    """Authorize a child/metadata write on `schema_id` and return the row.

    System tier → 403 (read-only). User tier → owner==caller else 404 (no
    existence oracle). Project tier → Manage grant on the project (resolve-to-
    owner). A missing schema → 404.
    """
    schema = await mutations.get_schema(schema_id)
    if schema is None:
        raise _not_found()
    if schema.scope == "system":
        raise _forbidden("system-tier schema is read-only")
    if schema.scope == "user":
        if schema.scope_id != str(caller):
            raise _not_found()  # not the owner — no existence oracle
        return schema
    # project tier — Manage-gate the project (resolve-to-owner).
    await _require_project_manage(schema.scope_id, caller, gc, projects)
    return schema


async def _require_project_manage(project_id: str | None, caller: UUID, gc, projects: ProjectsRepo) -> None:
    """Mirror require_project_grant(MANAGE) for a project_id pulled off a schema
    row (the grant dep keys off the path param; here the project id comes from
    the schema's scope_id)."""
    if project_id is None:
        raise _not_found()
    try:
        pid = UUID(project_id)
    except (ValueError, AttributeError):
        raise _not_found()
    meta = await projects.project_meta(pid)
    if meta is None:
        raise _not_found()
    owner, book_id = meta
    if caller == owner:
        return
    if book_id is None:
        raise _not_found()
    lvl = await gc.resolve_grant(book_id, caller)
    if lvl == GrantLevel.NONE:
        raise _not_found()
    if not lvl.at_least(GrantLevel.MANAGE):
        raise _forbidden()


def _tree_to_response(tree: dict) -> GraphSchemaTree:
    schema: GraphSchema = tree["schema"]
    return GraphSchemaTree(
        **schema.model_dump(),
        edge_types=[e.model_dump() for e in tree["edge_types"]],
        fact_types=[f.model_dump() for f in tree["fact_types"]],
        node_kinds=[k.model_dump() for k in tree["node_kinds"]],
        vocab_sets=[
            {
                **s.model_dump(),
                "values": [v.model_dump() for v in tree["vocab_values"].get(s.code, [])],
            }
            for s in tree["vocab_sets"]
        ],
    )


# ── list / read ───────────────────────────────────────────────────────────────
@router.get("/graph-schemas", response_model=GraphSchemaListResponse)
async def list_graph_schemas(
    scope: str | None = Query(default=None),
    project_id: str | None = Query(default=None),
    include_deprecated: bool = Query(default=False),
    user_id: UUID = Depends(get_current_user),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
) -> GraphSchemaListResponse:
    """List schemas visible to the caller: System (read-only) + the caller's
    User templates + (when `project_id` given) the project's adopted schema.

    Visibility is scope-keyed by the repo; the `project_id` filter only EXPOSES
    the caller's own project rows by id — it does not grant cross-project read."""
    items = await repo.list_visible(
        user_id, project_id=project_id, scope=scope, include_deprecated=include_deprecated
    )
    return GraphSchemaListResponse(items=items)


@router.get("/graph-schemas/{schema_id}", response_model=GraphSchemaTree)
async def get_graph_schema(
    schema_id: UUID = Path(),
    project_id: str | None = Query(default=None),
    include_deprecated: bool = Query(default=False),
    user_id: UUID = Depends(get_current_user),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
) -> GraphSchemaTree:
    """Read one schema with all children. Scope-visible to the caller (system to
    everyone; user to its owner; project to a caller passing its project_id)."""
    tree = await repo.get_tree(
        user_id, schema_id, project_id=project_id, include_deprecated=include_deprecated
    )
    if tree is None:
        raise _not_found()
    return _tree_to_response(tree)


@router.patch("/graph-schemas/{schema_id}", response_model=GraphSchema)
async def patch_graph_schema(
    body: GraphSchemaPatch,
    schema_id: UUID = Path(),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
) -> GraphSchema:
    """Edit user/project schema metadata. System tier → 403."""
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    return await mutations.patch_schema(
        schema_id,
        name=body.name,
        description=body.description,
        allow_free_edges=body.allow_free_edges,
    )


@router.delete("/graph-schemas/{schema_id}", status_code=status.HTTP_204_NO_CONTENT)
async def deprecate_graph_schema(
    schema_id: UUID = Path(),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
) -> None:
    """Soft-deprecate a user/project schema (recycle bin). System tier → 403."""
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    await mutations.deprecate_schema(schema_id)


# ── resolved schema ───────────────────────────────────────────────────────────
@router.get("/projects/{project_id}/schema")
async def get_resolved_schema(
    project_id: UUID = Path(),
    owner: UUID = Depends(require_project_grant(GrantLevel.VIEW)),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
):
    """The resolved effective schema for the project (system→user→project merge).
    View-gated on the project (resolve-to-owner)."""
    return await repo.resolve_for_project(str(project_id))


# ── adopt (copy-down, M1 adopt-gate) ──────────────────────────────────────────
@router.post(
    "/projects/{project_id}/adopt",
    response_model=GraphSchema,
    status_code=status.HTTP_201_CREATED,
)
async def adopt_schema(
    body: AdoptRequest,
    response: Response,
    project_id: UUID = Path(),
    owner: UUID = Depends(require_project_grant(GrantLevel.MANAGE)),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    projects: ProjectsRepo = Depends(get_projects_repo),
    glossary: GlossaryOntologyClient = Depends(get_glossary_ontology_client),
) -> GraphSchema:
    """Adopt (copy-down) a system/user template into the project. Manage-gated.

    M1 adopt-gate: resolve the project's glossary node-kind source (book ontology
    when the project has a book, else the owner's glossary standards), and block
    with 422 `needs_glossary` if the source template requires a node-kind the
    glossary doesn't have. Missing `optional` kinds proceed (warning header).
    Idempotent re-check: once the glossary gap is filled, re-adopt succeeds and
    replaces the project's active schema (one-active invariant).
    """
    project_str = str(project_id)
    # KM6-M2 + auto-seed: adopting a schema AUTO-CREATES the glossary node-kinds it
    # requires (copy-down from System), retrying the adopt once — shared with the
    # agent confirm effect so the GUI + MCP kg_adopt behave identically. A 422 now
    # only fires for a genuinely unsatisfiable gap (book-less project, or a required
    # kind with no System row to copy) — not for the common "glossary not seeded yet".
    try:
        result = await adopt_with_autocreate_glossary(
            projects, glossary, mutations,
            owner=owner, project_id=project_str, source_schema_id=body.source_schema_id,
        )
    except NeedsGlossaryError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={
                "code": "KG_ADOPT_NEEDS_GLOSSARY",
                "message": (
                    "the schema needs node-kinds that couldn't be auto-created (no "
                    "System kind to copy, or a book-less project) — add them in "
                    "glossary, then re-adopt"
                ),
                "needs_glossary": {"book_id": exc.book_id, "kinds": exc.kinds},
            },
        )
    except SchemaNotWritableError:
        raise _not_found()

    if result.missing_optional and not body.acknowledge_optional_gaps:
        response.headers["X-KG-Optional-Gaps"] = ",".join(result.missing_optional)
    return result.schema


# ── adopt loss preview (read-only, D-KG-LC-REVADOPT-LOSS) ──────────────────────
@router.post("/projects/{project_id}/adopt/preview")
async def adopt_preview(
    body: AdoptPreviewRequest,
    project_id: UUID = Path(),
    owner: UUID = Depends(require_project_grant(GrantLevel.MANAGE)),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
):
    """Preview "what you'll lose" before re-adopting a template. Read-only,
    Manage-gated (same grant as adopt — resolve-to-owner). Re-adopt deprecates
    the project's active schema and replaces it with a fresh copy of the source,
    silently dropping any customizations the source lacks; this surfaces them
    first so the UI can warn + gate the destructive action.

    Returns `{has_current, would_lose:[{node_type, code, change, ...}]}`. When the
    project never adopted (`has_current=False`) there is nothing to lose."""
    current_schema_id = await _active_project_schema_id(repo, str(project_id))
    try:
        return await mutations.compute_adopt_preview(
            owner_user_id=owner,
            project_id=str(project_id),
            current_schema_id=current_schema_id,
            incoming_source_id=body.source_schema_id,
        )
    except SchemaNotWritableError:
        raise _not_found()


# ── sync (tree diff / apply) ──────────────────────────────────────────────────
@router.get("/projects/{project_id}/sync/available")
async def sync_available(
    project_id: UUID = Path(),
    owner: UUID = Depends(require_project_grant(GrantLevel.VIEW)),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
):
    """Tree-granular diff of the project's active schema vs its upstream source.
    View-gated. Empty `changes` when up to date or the project never adopted."""
    schema_id = await _active_project_schema_id(repo, str(project_id))
    if schema_id is None:
        return {
            "source_ref": None,
            "source_hash_current": None,
            "project_source_hash": None,
            "has_updates": False,
            "changes": [],
        }
    return await mutations.sync_diff(schema_id)


@router.post("/projects/{project_id}/sync/apply")
async def sync_apply(
    body: SyncApplyRequest,
    project_id: UUID = Path(),
    owner: UUID = Depends(require_project_grant(GrantLevel.MANAGE)),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    repo: GraphSchemasRepo = Depends(get_graph_schemas_repo),
):
    """Apply per-node keep_mine/take_theirs. Manage-gated. Forward-only (M3).
    409 on `base_source_hash` drift (upstream moved since /available was read)."""
    schema_id = await _active_project_schema_id(repo, str(project_id))
    if schema_id is None:
        raise _not_found()
    try:
        return await mutations.sync_apply(
            schema_id,
            base_source_hash=body.base_source_hash,
            decisions=[d.model_dump() for d in body.decisions],
        )
    except SyncConflictError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="upstream source moved since the diff was read; re-fetch /sync/available",
        )
    except SchemaNotWritableError:
        raise _not_found()


async def _active_project_schema_id(repo: GraphSchemasRepo, project_id: str) -> UUID | None:
    """The current active (non-deprecated) project-scoped schema id, if any."""
    async with repo._pool.acquire() as conn:
        row = await conn.fetchrow(
            """
            SELECT schema_id FROM kg_graph_schemas
            WHERE scope = 'project' AND scope_id = $1 AND deprecated_at IS NULL
            ORDER BY updated_at DESC, schema_id DESC LIMIT 1
            """,
            project_id,
        )
    return row["schema_id"] if row else None


# ── child CRUD (additive + deprecate-only, M3) ───────────────────────────────
@router.post(
    "/graph-schemas/{schema_id}/edge-types",
    status_code=status.HTTP_201_CREATED,
)
async def add_edge_type(
    body: EdgeTypeCreate,
    schema_id: UUID = Path(),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
):
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    try:
        return await mutations.add_edge_type(
            schema_id,
            code=body.code,
            label=body.label,
            directed=body.directed,
            source_node_kinds=body.source_node_kinds,
            target_node_kinds=body.target_node_kinds,
            temporal=body.temporal,
            provenance_required=body.provenance_required,
            cardinality=body.cardinality,
            description=body.description,
        )
    except DuplicateChildError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="edge type code exists")


@router.delete(
    "/graph-schemas/{schema_id}/edge-types/{code}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def deprecate_edge_type(
    schema_id: UUID = Path(),
    code: str = Path(min_length=1, max_length=_CODE_MAX),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
) -> None:
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    try:
        await mutations.deprecate_edge_type(schema_id, code)
    except ChildNotFoundError:
        raise _not_found()


@router.post(
    "/graph-schemas/{schema_id}/fact-types",
    status_code=status.HTTP_201_CREATED,
)
async def add_fact_type(
    body: FactTypeCreate,
    schema_id: UUID = Path(),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
):
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    try:
        return await mutations.add_fact_type(
            schema_id, code=body.code, label=body.label, description=body.description
        )
    except DuplicateChildError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="fact type code exists")


@router.post(
    "/graph-schemas/{schema_id}/vocab-sets/{set_code}/values",
    status_code=status.HTTP_201_CREATED,
)
async def add_vocab_value(
    body: VocabValueCreate,
    schema_id: UUID = Path(),
    set_code: str = Path(min_length=1, max_length=_CODE_MAX),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
):
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    try:
        return await mutations.add_vocab_value(
            schema_id, set_code=set_code, code=body.code, label=body.label, metadata=body.metadata
        )
    except ChildNotFoundError:
        raise _not_found()
    except DuplicateChildError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="vocab value code exists")


@router.post(
    "/graph-schemas/{schema_id}/node-kinds",
    status_code=status.HTTP_201_CREATED,
)
async def add_node_kind(
    body: SchemaNodeKindCreate,
    schema_id: UUID = Path(),
    caller: UUID = Depends(get_current_user),
    mutations: OntologyMutationsRepo = Depends(get_ontology_mutations_repo),
    gc=Depends(get_grant_client),
    projects: ProjectsRepo = Depends(get_projects_repo),
):
    await _writable_schema_for_caller(schema_id, caller, mutations, gc, projects)
    try:
        return await mutations.add_node_kind(
            schema_id, kind_code=body.kind_code, strength=body.strength
        )
    except DuplicateChildError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="node kind exists")


# ── system-tier create (admin-only, requireAdmin placeholder) ─────────────────
@router.post("/system/graph-schemas")
async def create_system_schema() -> Response:
    """System-tier template create — admin-only. v1 returns 501 behind a
    `requireAdmin` placeholder until the admin-identity epic lands (contract
    freezes the route shape now)."""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="system-tier write is admin-only; not implemented in v1 (requireAdmin placeholder)",
    )

File: frontend/src/features/knowledge/types/ontology.ts
// ─────────────────────────────────────────────────────────────────────────────
// KG customizable-ontology types (lane LE).
// Mirrors contracts/api/knowledge-service/{ontology,views,triage}.yaml — the
// FROZEN S0/C0 shape. Keep field names verbatim with the contract so the
// mock→real API swap at C3 is a no-op for these types.
// ─────────────────────────────────────────────────────────────────────────────

export type Scope = 'system' | 'user' | 'project';
export type Strength = 'required' | 'optional';
export type Cardinality = 'single_active' | 'multi_active';

// ── graph schemas ────────────────────────────────────────────────────────────

export interface GraphSchemaSummary {
  schema_id: string;
  scope: Scope;
  scope_id?: string | null;
  code: string;
  name: string;
  description?: string | null;
  schema_version: number;
  /**
   * Q2 (LOCKED): true => off-vocab predicates allowed (today's behaviour);
   * false => closed to declared edge types (off-vocab → triage).
   */
  allow_free_edges: boolean;
  content_hash?: string | null;
  source_ref?: string | null;
  source_hash?: string | null;
  deprecated_at?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface EdgeType {
  edge_type_id?: string;
  code: string;
  label: string;
  directed: boolean;
  source_node_kinds?: string[];
  target_node_kinds?: string[];
  /** true => every instance requires valid_from + EVIDENCED_BY */
  temporal: boolean;
  provenance_required?: boolean;
  cardinality: Cardinality;
  description?: string | null;
  deprecated_at?: string | null;
}

export interface EdgeTypeCreate {
  code: string;
  label: string;
  directed?: boolean;
  source_node_kinds?: string[];
  target_node_kinds?: string[];
  temporal?: boolean;
  provenance_required?: boolean;
  cardinality?: Cardinality;
  description?: string;
}

export interface FactType {
  fact_type_id?: string;
  code: string;
  label: string;
  description?: string | null;
  deprecated_at?: string | null;
}

export interface FactTypeCreate {
  code: string;
  label: string;
  description?: string;
}

export interface VocabValue {
  vocab_value_id?: string;
  code: string;
  label: string;
  metadata?: Record<string, unknown>;
}

export interface VocabValueCreate {
  code: string;
  label: string;
  metadata?: Record<string, unknown>;
}

export interface VocabSet {
  vocab_set_id?: string;
  code: string;
  label: string;
  description?: string | null;
  /** true => extractor may only assign, not coin new values */
  closed: boolean;
  values?: VocabValue[];
}

export interface SchemaNodeKind {
  schema_node_kind_id?: string;
  kind_code: string;
  strength: Strength;
  deprecated_at?: string | null;
}

export interface SchemaNodeKindCreate {
  kind_code: string;
  strength: Strength;
}

export interface GraphSchemaTree extends GraphSchemaSummary {
  edge_types?: EdgeType[];
  fact_types?: FactType[];
  vocab_sets?: VocabSet[];
  /** BE returns vocab values SEPARATELY, keyed by vocab-set code (not nested in
   *  each set). `ontologyApi` nests them into `vocab_sets[].values` on read. */
  vocab_values?: Record<string, VocabValue[]>;
  node_kinds?: SchemaNodeKind[];
}

export interface GraphSchemaPatch {
  name?: string;
  description?: string;
  allow_free_edges?: boolean;
}

// ── full-CRUD authoring (A1/A2) — PATCH bodies are attribute-only; `code` is
//    IMMUTABLE (a rename = deprecate + create-new). ──────────────────────────
export interface EdgeTypePatch {
  label?: string;
  directed?: boolean;
  source_node_kinds?: string[];
  target_node_kinds?: string[];
  temporal?: boolean;
  provenance_required?: boolean;
  cardinality?: Cardinality;
  description?: string;
}

export interface FactTypePatch {
  label?: string;
  description?: string;
}

export interface NodeKindPatch {
  strength: Strength;
}

export interface VocabSetCreate {
  code: string;
  label: string;
  description?: string;
  closed?: boolean;
}

export interface VocabSetPatch {
  label?: string;
  description?: string;
  closed?: boolean;
}

export interface VocabValuePatch {
  label?: string;
  metadata?: Record<string, unknown>;
}

/** Create-from-scratch (A2) — a blank project schema (no template). */
export interface BlankSchemaCreate {
  name: string;
  description?: string;
  allow_free_edges?: boolean;
}

/** Clone a readable schema into a NEW user-scoped editable template (A2). */
export interface CloneSchemaRequest {
  source_schema_id: string;
  name?: string;
}

/** M3a — what the project's extracted graph already contains (to promote into the schema). */
export interface ObservedComponents {
  node_kinds: { code: string; count: number }[];
  edge_types: { code: string; count: number; source_kinds: string[]; target_kinds: string[] }[];
}

/** M3b — a schema proposal generated from a premise (LLM, propose→confirm). */
export interface SchemaProposeRequest {
  premise: string;
  genre?: string;
  model_ref: string;
}
export interface SchemaProposal {
  node_kinds: { code: string; label?: string }[];
  edge_types: { code: string; label?: string; source_kinds: string[]; target_kinds: string[] }[];
  fact_types: { code: string; label?: string }[];
}

export interface ResolvedSchema {
  project_id: string;
  schema_version: number;
  allow_free_edges: boolean;
  edge_types?: EdgeType[];
  fact_types?: FactType[];
  vocab_sets?: VocabSet[];
  /** Separate, keyed by vocab-set code (see GraphSchemaTree). Nested into
   *  `vocab_sets[].values` by `ontologyApi` on read. */
  vocab_values?: Record<string, VocabValue[]>;
  node_kinds?: SchemaNodeKind[];
}

// ── adopt ─────────────────────────────────────────────────────────────────────

export interface AdoptPayload {
  source_schema_id: string;
  /** proceed past missing `optional` kinds without re-warning */
  acknowledge_optional_gaps?: boolean;
}

/** Body of the M1 422 — adopt blocked on missing `required` glossary kinds. */
export interface NeedsGlossary {
  code: string;
  message: string;
  needs_glossary: {
    /** present when the project has a book; else kinds resolve vs user glossary standards */
    book_id?: string | null;
    /** the missing `required` node-kind codes the user must add in glossary first */
    kinds: string[];
  };
}

// ── adopt loss preview (re-adopt "what you'll lose") ──────────────────────────

export interface AdoptPreviewPayload {
  /** the candidate template to re-adopt; preview reports what the swap drops */
  source_schema_id: string;
}

/** One customization re-adopt would drop (removed_upstream) or overwrite (modified). */
export interface AdoptLoss {
  node_type: SyncNodeType;
  parent_code?: string | null;
  code: string;
  /** 'removed_upstream' = vanishes; 'modified' = overwritten by the template */
  change: 'removed_upstream' | 'modified';
  fields_changed?: string[];
}

export interface AdoptPreview {
  /** false => project never adopted; nothing to lose */
  has_current: boolean;
  would_lose: AdoptLoss[];
}

// ── sync ──────────────────────────────────────────────────────────────────────

export type SyncNodeType =
  | 'edge_type'
  | 'fact_type'
  | 'vocab_set'
  | 'vocab_value'
  | 'node_kind';

export type SyncChangeKind = 'added' | 'modified' | 'removed_upstream';

export interface SyncChange {
  node_type: SyncNodeType;
  parent_code?: string | null;
  code: string;
  change: SyncChangeKind;
  fields_changed?: string[];
  upstream?: Record<string, unknown> | null;
  mine?: Record<string, unknown> | null;
}

export interface SyncDiff {
  source_ref?: string | null;
  source_hash_current?: string | null;
  project_source_hash?: string | null;
  has_updates: boolean;
  changes: SyncChange[];
}

export type SyncChoice = 'keep_mine' | 'take_theirs';

export interface SyncDecision {
  node_type: SyncNodeType;
  parent_code?: string | null;
  code: string;
  choice: SyncChoice;
}

export interface SyncApplyPayload {
  /** the source_hash_current returned by /sync/available (optimistic-concurrency token) */
  base_source_hash: string;
  decisions: SyncDecision[];
}

export interface SyncApplyResult {
  schema_version: number;
  source_hash: string;
  applied: number;
}

// ── views ─────────────────────────────────────────────────────────────────────

export interface GraphView {
  view_id: string;
  project_id: string;
  user_id?: string;
  code: string;
  name: string;
  description?: string | null;
  edge_type_codes: string[];
  node_kind_codes: string[];
  created_at?: string;
  updated_at?: string;
}

export interface ViewCreate {
  /** slugified from name when omitted */
  code?: string;
  name: string;
  description?: string;
  edge_type_codes?: string[];
  node_kind_codes?: string[];
}

// ── graph read (temporal) ─────────────────────────────────────────────────────

export interface GraphNode {
  id: string;
  kind: string;
  name: string;
  glossary_entity_id?: string | null;
  // KG-ML M5 (C7) — localized labels for a reader whose language differs from
  // the source. null ⇒ keep the canonical kind/name (explicit source fallback).
  kind_label?: string | null;
  name_label?: string | null;
}

export interface GraphEdge {
  edge_type: string;
  source_id: string;
  target_id: string;
  valid_from?: number | null;
  valid_to?: number | null;
  schema_version?: number | null;
  // KG-ML M5 (C7) — localized predicate label (curated → humanized fallback).
  edge_type_label?: string | null;
}

export interface GraphSlice {
  as_of_chapter?: number | null;
  view?: string | null;
  nodes: GraphNode[];
  edges: GraphEdge[];
  warnings?: string[];
}

export interface GraphReadParams {
  /** view code; omitted = whole resolved schema */
  view?: string;
  /** chapter ordinal; omitted = latest (all open instances) */
  as_of_chapter?: number;
  limit?: number;
}

// ── triage ────────────────────────────────────────────────────────────────────

export type TriageItemType =
  | 'unknown_node_kind'
  | 'unknown_edge_type'
  | 'edge_kind_mismatch'
  | 'unknown_vocab_value'
  | 'edge_cardinality_conflict';

export type TriageStatus =
  | 'pending'
  | 'pending_glossary'
  | 'resolved'
  | 'dismissed';

export interface TriageGroup {
  signature: string;
  item_type: TriageItemType;
  count: number;
  status: TriageStatus;
  sample_payload?: Record<string, unknown>;
  suggested_actions?: string[];
}

export interface TriageGroupList {
  groups: TriageGroup[];
  next_cursor?: string | null;
}

// S-05 — per-item drill-in of a signature group (for single-item dismiss).
export interface TriageItem {
  triage_id: string;
  item_type: TriageItemType;
  payload?: Record<string, unknown>;
}

export interface TriageItemList {
  items: TriageItem[];
}

export interface TriageListParams {
  status?: TriageStatus;
  item_type?: TriageItemType;
  limit?: number;
  cursor?: string;
}

export type TriageAction =
  | 'map'
  | 'add_to_vocab'
  | 'add_to_schema'
  | 're_target'
  | 'widen_target_kinds'
  | 'drop_edge'
  | 'close_previous'
  | 'set_multi_active'
  | 'promote_to_glossary_kind'
  | 'demote_to_attribute'
  | 'dismiss';

export interface TriageResolvePayload {
  action: TriageAction;
  params?: Record<string, unknown>;
  /** when true, batch-apply to every pending item of this signature */
  apply_to_signature?: boolean;
}

export interface TriageResolveResult {
  status: 'resolved' | 'pending_glossary';
  affected: number;
  schema_version?: number | null;
  needs_glossary?: {
    book_id?: string | null;
    kinds: string[];
  } | null;
}
